pontoon/base/migrations/0133_calculate_strings_with_errors_warnings.py

```diff
@@ -0,0 +1,254 @@
+            else:
+                missing += 1
+
+        plural_unreviewed_count = translations.filter(
+            approved=False,
+            fuzzy=False,
+            rejected=False
+        ).count()
+        if plural_unreviewed_count:
+            unreviewed += 1
+
+    # Calculate diffs to reduce DB queries
+    total_strings_diff = resource.total_strings - self.total_strings
+    approved_strings_diff = approved - self.approved_strings
+    fuzzy_strings_diff = fuzzy - self.fuzzy_strings
+    strings_with_errors_diff = errors - self.strings_with_errors
+    strings_with_warnings_diff = warnings - self.strings_with_warnings
+    unreviewed_strings_diff = unreviewed - self.unreviewed_strings
+
+    # Translated Resource
+    adjust_stats(
+        self,
+        total_strings_diff,
+        approved_strings_diff,
+        fuzzy_strings_diff,
+        strings_with_errors_diff,
+        strings_with_warnings_diff,
+        unreviewed_strings_diff,
+    )
+
+    # Project
+    project = resource.project
+    adjust_stats(
+        project,
+        total_strings_diff,
+        approved_strings_diff,
+        fuzzy_strings_diff,
+        strings_with_errors_diff,
+        strings_with_warnings_diff,
+        unreviewed_strings_diff,
+    )
+
+    # Locale
+    if not project.system_project:
+        adjust_stats(
+            locale,
+            total_strings_diff,
+            approved_strings_diff,
+            fuzzy_strings_diff,
+            strings_with_errors_diff,
+            strings_with_warnings_diff,
+            unreviewed_strings_diff,
+        )
+
+    # ProjectLocale
+    project_locale = utils.get_object_or_none(
+        ProjectLocale,
+        project=project,
+        locale=locale
+    )
+    if project_locale:
+        adjust_stats(
+            project_locale,
+            total_strings_diff,
+            approved_strings_diff,
+            fuzzy_strings_diff,
+            strings_with_errors_diff,
+            strings_with_warnings_diff,
+            unreviewed_strings_diff,
+        )
+
+
+def calculate(apps, schema_editor):
+    """
+    Calculate `strings_with_errors` and `strings_with_warnings`.
+    """
+    TranslatedResource = apps.get_model('base', 'TranslatedResource')
+    Translation = apps.get_model('base', 'Translation')
+
+    Error = apps.get_model('checks', 'Error')
+    Warning = apps.get_model('checks', 'Warning')
+
+    # Collect all translations with errors and warnings.
+    translations = (
+        Translation.objects
+        .filter(
+            pk__in=(
+                list(Error.objects.values_list('translation', flat=True)) +
+                list(Warning.objects.values_list('translation', flat=True))
+            )
+        )
+        .prefetch_related('entity__resource')
+    )
+
+    # Collect TranslatedResources of translations with errors and warnings
+    # and run calculate_stats() on them.
+    query = Q()
+    for t in translations.values('entity__resource', 'locale', 'pk'):
```

Comment:
[nit] there's no reference to 'pk' later from what I see (also, with .distinct() you probably could reduce number of Q() objects)